### PR TITLE
editoast: handle consists changes in stdcm endpoint

### LINKED ISSUE #15163: editoast: handle consists changes in LMR
### Description and goal

editoast part of https://github.com/osrd-project/osrd-confidential/issues/809

### Acceptance criteria

- [ ] Adapt endpoint /stdcm
```rust
struct ConsistConfiguration {
   rolling_stock_id: i64,
   towed_rolling_stock_id: Option<i64>,
   total_mass: Option<quantities::Mass>,
   total_length: Option<quantities::Length>,
   max_speed: Option<quantities::Velocity>,
   speed_limit_tag: Option<String>,
   loading_gauge_type: Option<LoadingGaugeType>,
}
struct ConsistSchedule {
    boundaries: Vec<usize>, // Indices of steps where consist changes occur
    values: Vec<ConsistConfiguration>, // Consist configuration for each segment
}

pub(crate) struct Request {
   ...
   pub(crate) consist_schedule: ConsistSchedule,
   ...
}
```
boundaries should be strictly increasing and should not have 0 and step.len()
boundaries should have one less element than values

- [ ] Adapt VirtualTrainRun
For a consist change at point B, we should generate `ConsistConfiguration` and `VirtualTrain` A->B then B->C independatly. We retrieve `simulation_run_time()` of each `VirtualTrainRun` to compute `complete_simulation_run_time()`

- [ ] Adapt StdcmRequest 
```rust
pub struct ConsistConfiguration {
    pub rolling_stock_loading_gauge: LoadingGaugeType,
    pub rolling_stock_supported_signaling_systems: Vec<String>,
    pub speed_limit_tag: Option<String>,
    pub loading_gauge_type: Option<LoadingGaugeType>,
    #[schema(inline)]
    pub physics_consist: PhysicsConsist,
}

struct ConsistSchedule {
    boundaries: Vec<usize>, // Indices of steps where consist changes occur
    values: Vec<ConsistConfiguration>, // Consist configuration for each segment
}

pub struct Request {
   ...
   pub consist_schedule: ConsistSchedule,
   ...
}
```

### PR DESCRIPTION
**Pull request by ppitou**
Closes #15163 
For now, if no ConsistSchedule are given, it will use the current parameters (`rolling_stock_id` , `towed_rolling_stock_id`, ...)

### CODE REVIEW

File: editoast/src/views/timetable/stdcm.rs [hunk 1]
```diff
@@ -427,6 +458,51 @@ impl VirtualTrainRun {
...
+                app_version,
+                &request,
+                infra,
+                consist_parameters,
+            )
+            .await?;
+
+            let Some(simulation_run_time) = virtual_train_run.simulation.simulation_run_time().or_else(||
+                // Default 12h value when the simulation fails
+                matches!(virtual_train_run.simulation, simulation::Response::SimulationFailed { .. }).then_some(12 * 3_600_000))
```
Comment (MartinBourbier) [inline editoast/src/views/timetable/stdcm.rs]: Maybe introduce some kind of a constant?

**Review by MartinBourbier — COMMENTED** [review]
Thanks for the PR!

File: editoast/src/views/timetable/stdcm.rs [hunk 2]
```diff
@@ -19,6 +19,7 @@ use core_client::stdcm::Request as StdcmRequest;
 use core_client::stdcm::UndirectedTrackRange;
 use database::DbConnectionPoolV2;
 use editoast_derive::EditoastError;
+use itertools::Itertools;
```
Comment (leovalais) [inline editoast/src/views/timetable/stdcm.rs]:
```suggestion
use itertools::Itertools as _;
```

File: editoast/src/views/timetable/stdcm.rs [hunk 3]
```diff
@@ -242,73 +249,97 @@ pub(in crate::views) async fn stdcm_handler(
...
     };
 
+    for consist in &consist_configs {
+        let rolling_stock =
+            RollingStock::retrieve_or_fail(conn.clone(), consist.rolling_stock_id, || {
+                StdcmError::RollingStockNotFound {
+                    rolling_stock_id: consist.rolling_stock_id,
+                }
+            })
+            .await?
```
Comment (leovalais) [inline editoast/src/views/timetable/stdcm.rs]: nit: should be batched theoretically (though it's probably fine since an stdcm request won't realistically have a lot of consist changes), feel free to resolve

File: editoast/src/views/timetable/stdcm.rs [hunk 4]
```diff
@@ -125,6 +126,12 @@ pub(in crate::views) struct StdcmQueryParams {
     return_debug_payloads: bool,
 }
 
+#[allow(clippy::large_enum_variant)]
+enum ResultSimulation {
+    SimulationSuccess(u64),
+    SimulationError(simulation::Response),
+}
+
```
Comment (leovalais) [inline editoast/src/views/timetable/stdcm.rs]: My first reaction was "what's this?", "what is this u64?", etc. Naming is a bit sketchy and doesn't give more much information than a plain `Result<u64, simulation::Response>`.

Looking at where it's used and at `VirtualTrain`, I actually don't think this enum is necessary. More in following comments.

File: editoast/src/views/timetable/stdcm.rs [hunk 5]
```diff
@@ -427,6 +458,51 @@ impl VirtualTrainRun {
...
+    async fn simulate_consists_sequence(
+        db_pool: Arc<DbConnectionPoolV2>,
+        valkey_client: Arc<cache::Client>,
+        core_client: Arc<CoreClient>,
+        app_version: Option<&str>,
+        stdcm_request: &Request,
+        infra: &Infra,
+        consists_parameters: &[PhysicsConsistParameters],
+    ) -> Result<ResultSimulation> {
```
Comment (leovalais) [inline editoast/src/views/timetable/stdcm.rs]:
```suggestion
    ) -> Result<Vec<Self>> {
```

This is oddly asymmetric with `simulate`. Unless I'm mistaken, the postprocessing after each consist simulation has nothing to do with `VirtualTrain` and just anticipates the rest of the `stdcm_handler` workflow.

I'd suggest simplifying the VirtualTrain interface and lift the postprocessing one step higher. `ResultSimulation` would then have no reason to exist, which is also a plus imo.

File: editoast/src/views/timetable/stdcm.rs [hunk 6]
```diff
@@ -427,6 +458,51 @@ impl VirtualTrainRun {
...
+            let virtual_train_run = VirtualTrainRun::simulate(
+                db_pool.clone(),
+                valkey_client.clone(),
+                core_client.clone(),
+                app_version,
+                &request,
+                infra,
+                consist_parameters,
+            )
+            .await?;
```
Comment (leovalais) [inline editoast/src/views/timetable/stdcm.rs]: This should be batched as this results in multiple calls to `consist_train_simulation_batch` which does a lot internally.

File: editoast/src/views/timetable/stdcm.rs (hunk 1 above)
Comment (leovalais) [inline editoast/src/views/timetable/stdcm.rs]: nit: The double match on `virtual_train_run.simulation` is quite convoluted. A regular match would read better.

File: editoast/src/views/timetable/stdcm.rs [hunk 7]
```diff
@@ -427,6 +458,51 @@ impl VirtualTrainRun {
...
+
+        for ((start, end), consist_parameters) in itertools::chain!(
+            std::iter::once(0),
+            stdcm_request.consist_schedule.boundaries.clone(),
+            std::iter::once(stdcm_request.steps.len())
+        )
+        .tuple_windows()
+        .zip(consists_parameters.iter())
+        {
+            let mut request = stdcm_request.clone();
```
Comment (leovalais) [inline editoast/src/views/timetable/stdcm.rs]: note: the fact that we have to clone the whole request is a good indicator that we probably shouldn't rely on `simulate`.

File: editoast/src/views/timetable/stdcm.rs (hunk 1 above)
Comment (ppitou) [reply editoast/src/views/timetable/stdcm.rs]: Done

File: editoast/src/views/timetable/stdcm.rs (hunk 2 above)
Comment (ppitou) [reply editoast/src/views/timetable/stdcm.rs]: Done

File: editoast/src/views/timetable/stdcm.rs (hunk 3 above)
Comment (ppitou) [reply editoast/src/views/timetable/stdcm.rs]: it's batched now o7

File: editoast/src/views/timetable/stdcm.rs (hunk 4 above)
Comment (ppitou) [reply editoast/src/views/timetable/stdcm.rs]: Removed

File: editoast/src/views/timetable/stdcm.rs (hunk 1 above)
Comment (ppitou) [reply editoast/src/views/timetable/stdcm.rs]: I agree it was a bit confusing, done

File: editoast/src/views/timetable/stdcm.rs (hunk 6 above)
Comment (ppitou) [reply editoast/src/views/timetable/stdcm.rs]: ohhhh yes absolutely

File: editoast/src/views/timetable/stdcm.rs (hunk 5 above)
Comment (ppitou) [reply editoast/src/views/timetable/stdcm.rs]: Done

File: editoast/src/views/timetable/stdcm.rs [hunk 8]
```diff
@@ -242,54 +245,102 @@ pub(in crate::views) async fn stdcm_handler(
...
-            core_payload: None,
-        }))
-    };
+    let mut total_simulation_run_time = 0;
+    for virtual_train_run in &virtual_train_runs {
+        match virtual_train_run.simulation.simulation_run_time() {
+            Some(simulation_run_time) => total_simulation_run_time += simulation_run_time,
+            None => match virtual_train_run.simulation.clone() {
+                simulation::Response::SimulationFailed { .. } => {
+                    total_simulation_run_time += default_value_failed_simulation
```
Comment (aschantraine) [inline editoast/src/views/timetable/stdcm.rs]: I'm wondering whether we should simply set `total_simulation_run_time = default_value_failed_simulation` when the simulation fails. Right now, if two segments fail, we end up with 24h, whereas I believe the original 12h value was chosen to represent a full journey (before we introduced segmentation).

That said, I'm not entirely sure about the impact of having 12h vs 24h or more here, so it might not be a big deal.

File: front/public/locales/fr/errors.json [hunk 9]
```diff
@@ -226,12 +226,12 @@
       "UnknownSignalingSystem": "Système de signalisation inconnu"
     },
     "stdcm": {
+      "BatchRollingStockNotFound": "'{{count}}' matériel roulant(s) non trouvés",
```
Comment (aschantraine) [inline front/public/locales/fr/errors.json]:
```suggestion
      "BatchRollingStockNotFound": "'{{count}}' matériel roulant(s) non trouvé(s)",
```

**Review by aschantraine — COMMENTED** [review]
Thanks for this PR !

File: editoast/src/views/timetable/stdcm.rs [hunk 10]
```diff
@@ -79,13 +79,14 @@ pub(in crate::views) enum StdcmResponse {
...
+    #[error("Infrastructure {infra_id} does not exist")]
     InfraNotFound { infra_id: i64 },
     #[error("Timetable {timetable_id} does not exist")]
     #[editoast_error(status = 404)]
     TimetableNotFound { timetable_id: i64 },
-    #[error("Rolling stock {rolling_stock_id} does not exist")]
-    RollingStockNotFound { rolling_stock_id: i64 },
+    #[error("{count} rolling stock(s) could not be found")]
+    #[editoast_error(status = 404)]
+    BatchRollingStockNotFound { count: usize },
```
Comment (MartinBourbier) [inline editoast/src/views/timetable/stdcm.rs]: Maybe add the `ids` of the missing rolling stocks in this variant to better debug later

File: editoast/src/views/timetable/stdcm.rs [hunk 11]
```diff
@@ -298,26 +349,17 @@ pub(in crate::views) async fn stdcm_handler(
...
-                    .loading_gauge_type
-                    .unwrap_or(physics_consist_parameters.traction_engine.loading_gauge),
-                supported_signaling_systems: physics_consist_parameters
-                    .traction_engine
-                    .supported_signaling_systems(),
-                speed_limit_tag: request.speed_limit_tags.clone(),
-                physics_consist: physics_consist_parameters.into(),
-            }],
+            boundaries: request.consist_schedule.boundaries.clone(),
+            values: stdcm_consist_schedule_values,
```
Comment (MartinBourbier) [inline editoast/src/views/timetable/stdcm.rs:366]: I think there should be some sort of validation that `boundaries.len() + 1 == values.len()` since we rely on this invariant. I don't think I have seen it anywhere but I might be missing something

Comment (ppitou) [reply editoast/src/views/timetable/stdcm.rs:366]: It is already done [here](https://github.com/OpenRailAssociation/osrd/blob/dev/editoast/src/views/timetable/stdcm/request.rs#L422) or you want me to check again for core ?

File: editoast/src/views/timetable/stdcm.rs (hunk 10 above)
Comment (ppitou) [reply editoast/src/views/timetable/stdcm.rs]: Done

File: front/public/locales/fr/errors.json (hunk 9 above)
Comment (ppitou) [reply front/public/locales/fr/errors.json]: Thank you !

File: editoast/src/views/timetable/stdcm.rs (hunk 8 above)
Comment (ppitou) [reply editoast/src/views/timetable/stdcm.rs]: Oh yeah I didn't really know what was the best solution for here, if you think I should just set `total_simulation_run_time = default_value_failed_simulation` I'll change it

File: editoast/src/views/timetable/stdcm.rs [hunk 12]
```diff
@@ -218,6 +220,8 @@ pub(in crate::views) async fn stdcm_handler(
 
     let timetable_id = id;
     let infra_id = query.infra;
+    // Default 12h value when the simulation fails
+    let default_value_failed_simulation = 12 * 3_600_000;
```
Comment (leovalais) [inline editoast/src/views/timetable/stdcm.rs]:
```suggestion
    const DEFAULT_VALUE_FAILED_SIMULATION = 12 * 3_600_000;
```

File: editoast/src/views/timetable/stdcm.rs [hunk 13]
```diff
@@ -242,54 +246,100 @@ pub(in crate::views) async fn stdcm_handler(
...
     // 3. Get RollingStock
-    let rolling_stock =
-        RollingStock::retrieve_or_fail(conn.clone(), request.rolling_stock_id, || {
-            StdcmError::RollingStockNotFound {
-                rolling_stock_id: request.rolling_stock_id,
-            }
-        })
-        .await?
-        .into();
+    let mut physics_consists_parameters = vec![];
```
Comment (leovalais) [inline editoast/src/views/timetable/stdcm.rs]: nit: move the definition as close as possible from its first use, especially useful for reading since it's mutable.

File: editoast/src/views/timetable/stdcm.rs [hunk 14]
```diff
@@ -242,54 +246,100 @@ pub(in crate::views) async fn stdcm_handler(
...
-        max_speed: request.max_speed,
-        total_length: request.total_length,
-        total_mass: request.total_mass,
-        towed_rolling_stock,
-        traction_engine: rolling_stock,
-    };
+    let rolling_stocks: Vec<schemas::RollingStock> =
+        rolling_stocks_models.into_iter().map(Into::into).collect();
+
+    for (consist, rolling_stock) in consist_configs.iter().zip(rolling_stocks.into_iter()) {
```
Comment (leovalais) [inline editoast/src/views/timetable/stdcm.rs]:
```suggestion
    for (consist, rolling_stock) in consist_configs.iter().zip(rolling_stocks)) {
```

File: editoast/src/views/timetable/stdcm.rs (hunk 8 above)
Comment (leovalais) [reply editoast/src/views/timetable/stdcm.rs]: Good point, I guess it's fine to leave it like before.

File: editoast/src/views/timetable/stdcm.rs [hunk 15]
```diff
@@ -242,54 +246,100 @@ pub(in crate::views) async fn stdcm_handler(
...
-        // Default 12h value when the simulation fails
-        matches!(virtual_train_run.simulation, simulation::Response::SimulationFailed { .. }).then_some(12 * 3_600_000))
-    else {
-        return Ok(Json(StdcmResponse::PreprocessingSimulationError {
-            error: virtual_train_run.simulation,
-            core_payload: None,
-        }))
-    };
+    let mut total_simulation_run_time = 0;
+    for virtual_train_run in &virtual_train_runs {
```
Comment (leovalais) [inline editoast/src/views/timetable/stdcm.rs]:
```suggestion
    for virtual_train_run in virtual_train_runs {
```

Simulations aren't used after the loop, so doing like so prevents from cloning the entire simulation below.

---

There's still a double match in a loop mutating something, here's a simplification proposition (untested), wdty?

```diff
diff --git a/editoast/src/views/timetable/stdcm.rs b/editoast/src/views/timetable/stdcm.rs
index 27bad0506f..057dcc8a42 100644
--- a/editoast/src/views/timetable/stdcm.rs
+++ b/editoast/src/views/timetable/stdcm.rs
@@ -303,24 +304,26 @@ pub(in crate::views) async fn stdcm_handler(
     )
     .await?;
 
-    let mut total_simulation_run_time = 0;
-    for virtual_train_run in &virtual_train_runs {
-        match virtual_train_run.simulation.simulation_run_time() {
-            Some(simulation_run_time) => total_simulation_run_time += simulation_run_time,
-            None => match virtual_train_run.simulation.clone() {
-                simulation::Response::SimulationFailed { .. } => {
-                    total_simulation_run_time += default_value_failed_simulation
-                }
-                simulation => {
-                    return Ok(Json(StdcmResponse::PreprocessingSimulationError {
-                        error: simulation,
-                        core_payload: None,
-                    }));
-                }
-            },
-        }
+    let (mut pathfinding_failures, runs): (Vec<_>, Vec<_>) =
+        virtual_train_runs.into_iter().partition(|run| {
+            matches!(
+                run.simulation,
+                simulation::Response::PathfindingFailed { .. }
+            )
+        });
+    if let Some(failure) = pathfinding_failures.pop() {
+        return Ok(Json(StdcmResponse::PreprocessingSimulationError {
+            error: failure.simulation,
+            core_payload: None,
+        }));
     }
 
+    let total_simulation_run_time = runs
+        .iter()
+        .map(|run| run.simulation.simulation_run_time())
+        .sum::<Option<_>>()
+        .unwrap_or(default_value_failed_simulation);
+
     let earliest_departure_time = request.get_earliest_departure_time(total_simulation_run_time);
     let latest_simulation_end = request.get_latest_simulation_end(total_simulation_run_time); 
```

**Review by leovalais — COMMENTED** [review]
Thanks for the changes! We could use a test with a few consist changes.

File: editoast/src/views/timetable/stdcm.rs (hunk 8 above)
Comment (eckter) [reply editoast/src/views/timetable/stdcm.rs]: Higher value = longer processing time and possible timeouts. This gets significantly worse when the time is set at arrival and not departure, and it's not quite linear.

 Lower value = we may miss out on valid solutions.

Comment (eckter) [reply editoast/src/views/timetable/stdcm.rs]: Looking at the surrounding logic, there could be several solutions:

1. Don't add anything in the loop if it failed, set a flag, set the `total_simulation_run_time` to `max(total_simulation_run_time, 12h)` if one simulation failed
2.  Just set the total to 12h if any failed
3.  Introduce some kind of heuristic based on traveled distance, like 2h per 100km for the missing section
    * This require knowing the distance, an information that exists somewhere but may not be easily reachable here

Adding 12h at each failed simulation would be generally fine but could cause some issues in weird edge cases.

We could ask a PO to decide. I'd personally have a preference for 3 if realistic, 1 otherwise.

File: editoast/src/views/timetable/stdcm.rs (hunk 15 above)
Comment (leovalais) [reply editoast/src/views/timetable/stdcm.rs]: Note that the default value is incorrect, cf. https://github.com/OpenRailAssociation/osrd/pull/15969#discussion_r3032102977

File: editoast/src/views/timetable/stdcm.rs (hunk 12 above)
Comment (ppitou) [reply editoast/src/views/timetable/stdcm.rs]: Done

File: editoast/src/views/timetable/stdcm.rs (hunk 13 above)
Comment (ppitou) [reply editoast/src/views/timetable/stdcm.rs]: Noted! Done

File: editoast/src/views/timetable/stdcm.rs (hunk 14 above)
Comment (ppitou) [reply editoast/src/views/timetable/stdcm.rs]: Done

File: editoast/src/views/timetable/stdcm.rs (hunk 15 above)
Comment (ppitou) [reply editoast/src/views/timetable/stdcm.rs]: Woo yes much better thanks

File: editoast/src/views/timetable/stdcm.rs [hunk 16]
```diff
@@ -1095,4 +1144,167 @@ mod tests {
...
+        let third_rolling_stock =
+            create_fast_rolling_stock(&mut db_pool.get_ok(), &Uuid::new_v4().to_string()).await;
+
+        let mut payload = get_stdcm_payload(
+            first_rolling_stock.id,
+            None,
+            None,
+            None,
+            ConsistSchedule {
+                boundaries: vec![2, 4],
```
Comment (DucNg) [inline editoast/src/views/timetable/stdcm.rs]: Could you add a test where the boundaries is [1]?

File: editoast/src/views/timetable/stdcm.rs [hunk 17]
```diff
@@ -1095,4 +1144,167 @@ mod tests {
...
+            create_fast_rolling_stock(&mut db_pool.get_ok(), &Uuid::new_v4().to_string()).await;
+        let second_rolling_stock =
+            create_fast_rolling_stock(&mut db_pool.get_ok(), &Uuid::new_v4().to_string()).await;
+        let third_rolling_stock =
+            create_fast_rolling_stock(&mut db_pool.get_ok(), &Uuid::new_v4().to_string()).await;
+
+        let mut payload = get_stdcm_payload(
+            first_rolling_stock.id,
+            None,
+            None,
```
Comment (DucNg) [inline editoast/src/views/timetable/stdcm.rs:1364]: Can we try adding the total_mass?

**Review by DucNg — COMMENTED** [review]
Could you try adding more tests cases?

File: editoast/src/views/timetable/stdcm.rs (hunk 16 above)
Comment (ppitou) [reply editoast/src/views/timetable/stdcm.rs]: Done

File: editoast/src/views/timetable/stdcm.rs (hunk 17 above)
Comment (ppitou) [reply editoast/src/views/timetable/stdcm.rs:1364]: Done

File: editoast/src/views/timetable/stdcm/request.rs [hunk 18]
```diff
@@ -58,7 +58,7 @@ pub(crate) struct ConsistConfiguration {
     pub(crate) loading_gauge_type: Option<LoadingGaugeType>,
 }
 
-#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, ToSchema)]
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, ToSchema, Default)]
```
Comment (leovalais) [inline editoast/src/views/timetable/stdcm/request.rs]:
```suggestion
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, ToSchema)]
#[cfg_attr(test, derive(Default))]
```

If it's only for tests better make it explicit as the default values for such a structure isn't obvious (to me at least).

File: editoast/src/views/timetable/stdcm.rs [hunk 19]
```diff
@@ -1039,4 +1148,256 @@ mod tests {
         // THEN
         assert!(filtered.is_empty() == filtered_out);
     }
+
+    #[tokio::test(flavor = "multi_thread", worker_threads = 1)]
```
Comment (leovalais) [inline editoast/src/views/timetable/stdcm.rs]:
```suggestion
    #[test]
```

just use test if you don't need async

File: editoast/src/views/timetable/stdcm.rs [hunk 20]
```diff
@@ -1039,4 +1148,256 @@ mod tests {
...
+            .json(&payload);
+
+        let stdcm_response: StdcmResponse = app
+            .fetch(request)
+            .await
+            .assert_status(StatusCode::OK)
+            .json_into();
+
+        if let PathfindingResult::Success(path) =
+            PathfindingResult::Success(pathfinding_result_success())
```
Comment (leovalais) [inline editoast/src/views/timetable/stdcm.rs]: ?

**Review by leovalais — APPROVED** [review]
LGTM, thanks for the changes!

File: editoast/src/views/timetable/stdcm.rs (hunk 15 above)
Comment (leovalais) [reply editoast/src/views/timetable/stdcm.rs]: The default value is still 12h, @eckter is it ok?

File: editoast/src/views/timetable/stdcm/request.rs (hunk 18 above)
Comment (ppitou) [reply editoast/src/views/timetable/stdcm/request.rs]: Done

File: editoast/src/views/timetable/stdcm.rs (hunk 19 above)
Comment (ppitou) [reply editoast/src/views/timetable/stdcm.rs]: Oh yes right

File: editoast/src/views/timetable/stdcm.rs (hunk 20 above)
Comment (ppitou) [reply editoast/src/views/timetable/stdcm.rs]: 🗿

File: editoast/src/views/timetable/stdcm.rs (hunk 8 above)
Comment (ppitou) [reply editoast/src/views/timetable/stdcm.rs]: After discussion, we decided to stay the way we are for now

**Review by SharglutDev — APPROVED** [review]
Lgtm for front ✅ (not tested)

**Review by DucNg — APPROVED** [review]
Lgtm for front ✅ (**and tested :rocket: !!**)
